### PR TITLE
feat: impl r1cs cache in prover (#71)

### PR DESCRIPTION
**Pull request by captainlee1024**
feat: impl r1cs cache in prover (#71), optimize proof speed

### CODE REVIEW

File: prover/src/provers/agg_prover.rs [hunk 1]
```diff
@@ -78,21 +98,46 @@ impl Prover<AggContext> for AggProver {
...
+                &r1_stark.const_file,
+                &r1_stark.exec_file,
+                force_bits,
+            )?;
+
+            // add r1cs pil, const, exec to cache and update flag
+            let _ = prove_data_cache.add(
+                r1_stark.r1cs_file.clone(),
+                CacheStage::Agg(StarkFileType::R1cs),
+            );
```
Comment (eigmax) [inline prover/src/provers/agg_prover.rs:115]: Use add(xx)? to catch the errors.

File: prover/src/provers/agg_prover.rs [hunk 2]
```diff
@@ -78,21 +98,46 @@ impl Prover<AggContext> for AggProver {
...
 
         // 4. compress exec
         // TODO: place it in StarkProveArgs?
         let wasm_file = format!(
             "{}/{}.recursive1_js/{}.recursive1.wasm",
             r1_circom.output, ctx.task_name, ctx.task_name
         );
+
         log::info!("wasm_file: {}", wasm_file);
+        log::info!("exec");
```
Comment (eigmax) [inline prover/src/provers/agg_prover.rs:139]: Is this logging useful?

File: prover/src/provers/final_prover.rs [hunk 3]
```diff
@@ -19,35 +20,63 @@ impl FinalProver {
...
+        if prove_data_cache.final_cache.already_cached {
+            r2.r1cs_file
+                .clone_from(&prove_data_cache.final_cache.r1cs_file);
+            r2.pil_file
+                .clone_from(&prove_data_cache.final_cache.pil_file);
+            r2.const_file
+                .clone_from(&prove_data_cache.final_cache.const_file);
+            r2.exec_file
+                .clone_from(&prove_data_cache.final_cache.exec_file);
+        }
```
Comment (eigmax) [inline prover/src/provers/final_prover.rs:43]: here is a classical if else expression, just use if else.

File: prover/src/provers/agg_prover.rs (hunk 1 above)
Comment (captainlee1024) [reply prover/src/provers/agg_prover.rs]: done

File: prover/src/provers/agg_prover.rs (hunk 2 above)
Comment (captainlee1024) [reply prover/src/provers/agg_prover.rs]: I think it should be removed too. already done!

**Review by JangoCCC — APPROVED** [review]

File: prover/src/provers/final_prover.rs (hunk 3 above)
Comment (captainlee1024) [reply prover/src/provers/final_prover.rs]: ok, already optimized

**Pull request merged**